### PR TITLE
feat(tools): let toolsets replace a failed tools listing

### LINKED ISSUE #6748: Applications cannot influence what a toolset contributes when tool listing fails (per-user OAuth MCP servers)
**Is your feature request related to a problem? Please describe.**

The context is per-user OAuth (3LO) MCP servers: each end user must individually
authorize the remote server, and until they do, McpToolset tool listing fails with
HTTP 401. This is not an outage — it is the expected steady state for every user who
has not connected yet, and the product requirement is to prompt that user to connect.

Since 2.7.0, SkillToolset._resolve_additional_tools_from_state gathers with
return_exceptions=True and skips a failing toolset with a warning. That isolation is
right for what it targets: one unreachable server must not take down unrelated tools.
The side effect for the 3LO case is that the failing toolset's own tools disappear
silently. Skill adk_additional_tools are resolved by name against the candidate pool,
so the names served by that MCP server never enter the pool. The model therefore has
no tool to call, never observes an authorization error, and has no way to ask the user
to authorize. The user is simply told the capability does not exist.

#4300 was closed with the guidance to wrap the get_tools() call in try/except
ConnectionError. That works when the application calls get_tools itself. It does not
when the toolset is owned by an agent or by SkillToolset, because the framework makes
the call. The only remaining seam is for the application to wrap every MCP toolset in
its own BaseToolset subclass, which is per-toolset opt-in and fails silently the day
someone adds a new toolset and forgets.

**Describe the solution you'd like**

An extension point that lets the owning application decide what a toolset contributes
when listing fails, instead of the framework deciding it always contributes nothing.
Conceptually an overridable hook on BaseToolset, invoked with the raised exception and
the readonly context, whose return value replaces that toolset's tool list for the
current resolution, with a default that behaves exactly as today (skip and warn). An
application could then return placeholder tools carrying an "authorization required"
instruction, so the authorization prompt reaches the user through the normal
tool-calling path.

**Describe alternatives you've considered**

Wrapping each MCP toolset in an application-side BaseToolset. This is what we do today;
it works, but it is opt-in per toolset and silent when omitted.

Checking authorization before building the toolset. This duplicates the server's own
auth decision and races with token expiry between the check and the listing.

Special-casing 401 inside McpToolset. Too opinionated for the framework — what should
happen on an unauthorized listing is application- and product-specific.

**Additional context**

Observed on google-adk 2.7.0. This affects any ADK application fronting MCP servers
that authorize per end user rather than per application.

---

Not a Contribution: this issue is submitted solely to report a limitation and request a
feature. It is expressly not intended as a Contribution under the Google CLA, and no
code, patch, or other work of authorship is submitted with it.

### PR DESCRIPTION
**Pull request by a2105z**
## Summary
- Add `BaseToolset.on_tools_listing_error` so apps can contribute replacement tools when listing fails (default: empty list / current skip behavior)
- Invoke the hook from `LlmAgent` tool loading and `SkillToolset` additional-tool resolution after the existing isolate-and-continue logging
- Enables per-user OAuth (3LO) MCP flows to surface connect/authorize placeholder tools instead of silently dropping the toolset

Fixes #6748

## Test plan
- [x] `pytest tests/unittests/tools/test_base_toolset.py` (default + override hook)
- [x] `pytest tests/unittests/agents/test_llm_agent_fields.py -k canonical_tools`
- [x] `pytest tests/unittests/tools/test_skill_toolset.py -k resolution_`
- [ ] Reviewer: confirm hook is not invoked on direct `get_tools()` / `get_tools_with_prefix()` calls (exceptions still propagate there)

### CODE REVIEW

**Comment by a2105z** [comment]
@googlebot I fixed it.

File: src/google/adk/tools/base_toolset.py [hunk 1]
```diff
@@ -99,6 +99,35 @@ async def get_tools(
...
+    Direct callers of ``get_tools`` / ``get_tools_with_prefix`` still see the
+    original exception; this hook is only invoked by the framework's
+    isolate-and-continue paths.
+
+    Args:
+      error: The exception raised by ``get_tools`` / ``get_tools_with_prefix``.
+      readonly_context: The same context passed to the failed listing call.
+
+    Returns:
+      Tools to use in place of the failed listing. The default returns an
```
Comment (varunbiluri) [inline src/google/adk/tools/base_toolset.py:129]: The replacement list bypasses `get_tools_with_prefix`, so a toolset configured with `tool_name_prefix` returns unprefixed fallback tools. That breaks the BaseToolset naming contract precisely on the failure path and can introduce collisions with other toolsets (the canonical-tools caller appends this list directly, and SkillToolset does likewise). Please apply the same prefix transformation to fallback tools, or explicitly route hook results through a shared prefix helper, and add a regression test using a non-empty prefix.

**Review by varunbiluri — COMMENTED** [review]
Staff review: one prefix-contract issue to address.

File: src/google/adk/tools/base_toolset.py (hunk 1 above)
Comment (a2105z) [reply src/google/adk/tools/base_toolset.py:129]: Thanks @varunbiluri — great catch.

Fixed by extracting `_apply_tool_name_prefix` and routing listing-error fallbacks through it from both `LlmAgent.canonical_tools` and `SkillToolset` additional-tool resolution (same transformation as `get_tools_with_prefix`). Hook overrides still return unprefixed names; the framework applies the prefix.

Added regression coverage for a non-empty `tool_name_prefix` on the agent path, base toolset helper, and skill additional-tools path.

Appreciate the staff review — anything else you’d like me to tighten up here?

**Comment by a2105z** [comment]
@varunbiluri addressed your prefix-contract note in the latest push — thank you again for the staff review. Please take another look when you have a chance, and let me know if there’s anything else you’d like changed.

**Review by varunbiluri — APPROVED** [review]
Re-reviewed `08bcf768`. Fallback tools now go through the same copy-and-prefix transformation as successful listings in both the agent and SkillToolset isolation paths, with behavioral coverage at each boundary. The hook contract clearly tells overrides to return unprefixed names. My prefix-contract concern is resolved.

**Comment by a2105z** [comment]
@varunbiluri  Thank you for the review - hopefully we can get this merged soon then.
